### PR TITLE
Implement new getPegoutCreationRskTxHashByBtcTxHash to fetch pegout creation index

### PR DESCRIPTION
**Pull request by nathanieliov**
This improvement adds the new getPegoutCreationRskTxHashByBtcTxHash bridge method to federatorSupport and then implements its use on the BtcReleaseClient (tryGetReleaseInformation method) in order to fetch the peg-out rskTxHash from the index, if it is not found there, it fetches the peg-out rskTxHash from the legacy storage.

### CODE REVIEW

**Comment by nathanieliov** [comment]
pipeline:run

File: src/test/java/co/rsk/federate/FederatorSupportTest.java [hunk 1]
```diff
@@ -107,7 +117,56 @@ public void hasBlockCoinbaseInformed() {
         }).when(bridgeTransactionSender).callTx(any(), any(), any());
 
         Assert.assertTrue(fs.hasBlockCoinbaseInformed(Sha256Hash.ZERO_HASH));
+    }
+
+    @Test
+    public void getPegoutCreationRskTxHashByBtcTxHash() {
```
Comment (jeremy-then) [inline src/test/java/co/rsk/federate/FederatorSupportTest.java:123]: Interesting. Why is this a `get` method but does not return anything?

**Review by jeremy-then — COMMENTED** [review]
LGTM.

**Comment by kelvinator07** [comment]
BtcReleaseClientService Has Only One Implementation, So Using The BtcReleaseClientService Class Should Be Proper Instead Of BtcReleaseClientServiceImpl... What Do You Think?

File: src/test/java/co/rsk/federate/btcreleaseclient/BtcReleaseClientServiceImplTest.java [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+        test_getRskTxHash( sha256Hash, hash, false, false);
+    }
+
+    private void test_getRskTxHash(Sha256Hash sha256Hash,
+                                   Keccak256 hash,
+                                   boolean fromPegoutCreationIndex,
+                                   boolean fromLegacyStorage) {
+        BridgeTransactionSender bridgeTransactionSender = mock(BridgeTransactionSender.class);
+
+        FederatorSupport fs = Mockito.spy(new FederatorSupport(
```
Comment (kelvinator07) [inline src/test/java/co/rsk/federate/btcreleaseclient/BtcReleaseClientServiceImplTest.java:135]:
```suggestion
        FederatorSupport federatorSupport = Mockito.spy(new FederatorSupport(
```

File: src/test/java/co/rsk/federate/btcreleaseclient/BtcReleaseClientServiceImplTest.java [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+    public void hasBtcTxHash_not_found() {
+        Sha256Hash sha256Hash = TestUtils.createBtcTransaction(params, federation).getHash();
+        Keccak256 hash = TestUtils.createHash(25);
+        hasBtcTxHash( sha256Hash, hash, false, false);
+    }
+
+    private void hasBtcTxHash(Sha256Hash sha256Hash, Keccak256 hash, boolean fromPegoutCreationIndex, boolean fromLegacyStorage) {
+        BridgeTransactionSender bridgeTransactionSender = mock(BridgeTransactionSender.class);
+
+        FederatorSupport fs = Mockito.spy(new FederatorSupport(
```
Comment (kelvinator07) [inline src/test/java/co/rsk/federate/btcreleaseclient/BtcReleaseClientServiceImplTest.java:66]:
```suggestion
        FederatorSupport federatorSupport = Mockito.spy(new FederatorSupport(
```

File: src/test/java/co/rsk/federate/FederatorSupportTest.java [hunk 4]
```diff
@@ -107,7 +117,56 @@ public void hasBlockCoinbaseInformed() {
...
+        co.rsk.bitcoinj.core.Sha256Hash sha256Hash = co.rsk.bitcoinj.core.Sha256Hash.ZERO_HASH;
+        Keccak256 hash = Keccak256.ZERO_HASH;
+
+        test_getPegoutCreationRskTxHashByBtcTxHash(sha256Hash, hash);
+    }
+
+    private void test_getPegoutCreationRskTxHashByBtcTxHash(co.rsk.bitcoinj.core.Sha256Hash sha256Hash, Keccak256 keccak256) {
+        BridgeTransactionSender bridgeTransactionSender = mock(BridgeTransactionSender.class);
+
+        FederatorSupport fs = new FederatorSupport(
```
Comment (kelvinator07) [inline src/test/java/co/rsk/federate/FederatorSupportTest.java:144]:
```suggestion
        FederatorSupport federatorSupport = new FederatorSupport(
```

**Comment by nathanieliov** [comment]
> BtcReleaseClientService Has Only One Implementation, So Using The BtcReleaseClientService Class Should Be Proper Instead Of BtcReleaseClientServiceImpl... What Do You Think?

@kelvinator07 The idea is to reference the interface instead of a concrete class, to abstract the consumers from a specific implementation. This way, if later we want to use a BtcReleaseClientServiceImplV2 we just need to implement the. BtcReleaseClientService interface, then build the v2 object instead.

File: src/test/java/co/rsk/federate/btcreleaseclient/BtcReleaseClientServiceImplTest.java (hunk 2 above)
Comment (nathanieliov) [reply src/test/java/co/rsk/federate/btcreleaseclient/BtcReleaseClientServiceImplTest.java]: Done

File: src/test/java/co/rsk/federate/btcreleaseclient/BtcReleaseClientServiceImplTest.java (hunk 3 above)
Comment (nathanieliov) [reply src/test/java/co/rsk/federate/btcreleaseclient/BtcReleaseClientServiceImplTest.java]: Done

File: src/test/java/co/rsk/federate/FederatorSupportTest.java (hunk 4 above)
Comment (nathanieliov) [reply src/test/java/co/rsk/federate/FederatorSupportTest.java]: Done

**Review by jeremy-then — APPROVED** [review]
LGTM.

File: src/test/java/co/rsk/federate/btcreleaseclient/BtcReleaseClientServiceImplTest.java [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+    public void hasBtcTxHash_not_found() {
+        Sha256Hash sha256Hash = TestUtils.createBtcTransaction(params, federation).getHash();
+        Keccak256 hash = TestUtils.createHash(25);
+        hasBtcTxHash( sha256Hash, hash, false, false);
+    }
+
+    private void hasBtcTxHash(Sha256Hash sha256Hash, Keccak256 hash, boolean fromPegoutCreationIndex, boolean fromLegacyStorage) {
+        BridgeTransactionSender bridgeTransactionSender = mock(BridgeTransactionSender.class);
+
+        FederatorSupport federatorSupport = Mockito.spy(new FederatorSupport(
```
Comment (marcos-iov) [inline src/test/java/co/rsk/federate/btcreleaseclient/BtcReleaseClientServiceImplTest.java:66]: I would simply mock the FederatorSupport to return whatever response we expect. The purpose here is to test BtcReleaseClientServiceImpl class, not FederatorSupport

File: src/main/java/co/rsk/federate/btcreleaseclient/BtcReleaseClientService.java [hunk 6]
```diff
@@ -0,0 +1,11 @@
+package co.rsk.federate.btcreleaseclient;
+
+import co.rsk.bitcoinj.core.Sha256Hash;
+import co.rsk.crypto.Keccak256;
+
+import java.util.Optional;
+
+public interface BtcReleaseClientService {
+    boolean hasBtcTxHash(Sha256Hash btcTxHash);
+    Optional<Keccak256> getRskTxHash(Sha256Hash btcTxHash);
```
[the remark this answers is not in the document]
Comment (nathanieliov) [reply src/main/java/co/rsk/federate/btcreleaseclient/BtcReleaseClientService.java:10]: Regarding the first comment, good idea, we may remove the `hasBtcTxHash` method and use the Optional to know if the`rskTxHash` was found or not.

File: src/test/java/co/rsk/federate/FederatorSupportTest.java [hunk 7]
```diff
@@ -15,13 +21,17 @@
...
 
 import static org.mockito.Mockito.*;
 
 public class FederatorSupportTest {
 
     private NetworkParameters networkParameters = ThinConverter.toOriginalInstance(BridgeRegTestConstants.getInstance().getBtcParamsString());
 
+    private co.rsk.bitcoinj.core.NetworkParameters params = RegTestParams.get();
+    private Federation federation = TestUtils.createFederation(params, 1);
+
```
[the remark this answers is not in the document]
Comment (nathanieliov) [reply src/test/java/co/rsk/federate/FederatorSupportTest.java:34]: What's the different on doing in before method or just doing it when the class got instantiate?

File: src/test/java/co/rsk/federate/FederatorSupportTest.java [hunk 8]
```diff
@@ -107,7 +117,56 @@ public void hasBlockCoinbaseInformed() {
         }).when(bridgeTransactionSender).callTx(any(), any(), any());
 
         Assert.assertTrue(fs.hasBlockCoinbaseInformed(Sha256Hash.ZERO_HASH));
+    }
+
+    @Test
+    public void getPegoutCreationRskTxHashByBtcTxHash() {
+        co.rsk.bitcoinj.core.NetworkParameters params = RegTestParams.get();
+        Federation federation = TestUtils.createFederation(params, 1);
```
[the remark this answers is not in the document]
Comment (nathanieliov) [reply src/test/java/co/rsk/federate/FederatorSupportTest.java:125]: Good catch

File: src/main/java/co/rsk/federate/btcreleaseclient/BtcReleaseClientService.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (marcos-iov) [reply src/main/java/co/rsk/federate/btcreleaseclient/BtcReleaseClientService.java]: Did you forget to change the method name?

File: src/test/java/co/rsk/federate/btcreleaseclient/BtcReleaseClientServiceImplTest.java [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+        test_getRskTxHash( sha256Hash, hash, false, false);
+    }
+
+    private void test_getRskTxHash(Sha256Hash sha256Hash,
+                                   Keccak256 hash,
+                                   boolean fromPegoutCreationIndex,
+                                   boolean fromLegacyStorage) {
+        BridgeTransactionSender bridgeTransactionSender = mock(BridgeTransactionSender.class);
+
+        FederatorSupport federatorSupport = Mockito.spy(new FederatorSupport(
```
Comment (marcos-iov) [inline src/test/java/co/rsk/federate/btcreleaseclient/BtcReleaseClientServiceImplTest.java:62]: What do you think of simply mocking FederatorSupport? The call to the Bridge is already tested in FederatorSupportTest.

Here you simplify to something like this
```
when(federatorSupport.getPegoutCreationRskTxHashByBtcTxHash).thenReturn(...)
```

File: src/test/java/co/rsk/federate/FederatorSupportTest.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (marcos-iov) [reply src/test/java/co/rsk/federate/FederatorSupportTest.java]: Looks like there is none after all, https://stackoverflow.com/questions/1994218/should-i-instantiate-instance-variables-on-declaration-or-in-the-constructor

So, feel free to use whatever approach you like

File: src/test/java/co/rsk/federate/FederatorSupportTest.java [hunk 10]
```diff
@@ -1,26 +1,46 @@
...
 import co.rsk.peg.BridgeMethods;
+import co.rsk.peg.Federation;
+import co.rsk.peg.FederationMember;
 import org.bitcoinj.core.*;
 import org.bouncycastle.util.encoders.Hex;
 import org.ethereum.core.Blockchain;
+import org.ethereum.util.ByteUtil;
 import org.junit.Assert;
+import org.junit.Before;
+import org.junit.BeforeClass;
```
Comment (marcos-iov) [inline src/test/java/co/rsk/federate/FederatorSupportTest.java:20]: Is this used?

File: src/test/java/co/rsk/federate/FederatorSupportTest.java [hunk 11]
```diff
@@ -1,26 +1,46 @@
...
+    private NetworkParameters networkParameters;
+
+    private co.rsk.bitcoinj.core.NetworkParameters params;
+    private Federation federation;
+
+    @Before
+    public void setUp() throws Exception {
+        networkParameters = ThinConverter.toOriginalInstance(BridgeRegTestConstants.getInstance().getBtcParamsString());
+        params = RegTestParams.get();
+        federation = TestUtils.createFederation(params, 1);
```
Comment (marcos-iov) [inline src/test/java/co/rsk/federate/FederatorSupportTest.java:42]:
```suggestion
        federation = TestUtils.createFederation(RegTestParams.get(), 1);
```
No need to create the class variable `params` if it's only used here

File: src/test/java/co/rsk/federate/FederatorSupportTest.java [hunk 12]
```diff
@@ -1,26 +1,46 @@
...
 public class FederatorSupportTest {
 
-    private NetworkParameters networkParameters = ThinConverter.toOriginalInstance(BridgeRegTestConstants.getInstance().getBtcParamsString());
+    private NetworkParameters networkParameters;
+
+    private co.rsk.bitcoinj.core.NetworkParameters params;
+    private Federation federation;
+
+    @Before
+    public void setUp() throws Exception {
```
Comment (marcos-iov) [inline src/test/java/co/rsk/federate/FederatorSupportTest.java:39]:
```suggestion
    public void setUp() {
```
I don't think it's throwing any exception

File: src/main/java/co/rsk/federate/btcreleaseclient/BtcReleaseClientService.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (nathanieliov) [reply src/main/java/co/rsk/federate/btcreleaseclient/BtcReleaseClientService.java]: 🤦‍♂️ Yes, I did. I'll resolve it.

File: src/test/java/co/rsk/federate/FederatorSupportTest.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (nathanieliov) [reply src/test/java/co/rsk/federate/FederatorSupportTest.java]: Cool. Thanks for sharing. Useful reading. Knowing now that this is added by the compiler at the end of the constructor block, I think can be useful putting it into a before method that will force those variables to always be initialized.

File: src/test/java/co/rsk/federate/FederatorSupportTest.java (hunk 11 above)
Comment (nathanieliov) [reply src/test/java/co/rsk/federate/FederatorSupportTest.java:42]: It is being used in more places. For ex:


image

Comment (marcos-iov) [reply src/test/java/co/rsk/federate/FederatorSupportTest.java:40]: I see, thanks

**Review by marcos-iov — APPROVED** [review]
Great work!

**Pull request merged**